### PR TITLE
Multivariate regression tutorial

### PR DESCRIPTION
**Pull request by codingS3b**
Based on our conversation [here](https://github.com/lightning-uq-box/lightning-uq-box/pull/148#issuecomment-2144791943), I polished a notebook that demonstrates how to use the existing deep-ensemble functionality of the toolbox to perform a multivariate regression problem of a 1D input signal to a 1D output signal.

### CODE REVIEW

**Comment by nilsleh** [comment]
Very cool, thanks a lot for that contribution! I think I only have two comments:

- formatting, should work with `pip install ruff` and then `ruff format` 
- maybe adding another intro paragraph in the notebook in the beginning, saying explicitly that this is about multi-variate regression at the very start and also maybe linking to https://github.com/lightning-uq-box/lightning-uq-box/pull/148 to say that one could currently do multi-variate regression like demonstrated in the notebook, but the implementation might be subject to change potentially?

**Comment by codingS3b** [comment]
Thanks for your feedback. I tried including that in the latest commits. However, despite running `ruff format` before committing, I'm now still getting ruff errors complaining about import ordering which I find strange as I think they are already sorted alphabetically (maybe the problem is dividing between `import ...` and `from ... import ...`?

@elcorto and I also looked a bit deeper into the calibration behavior and distribution of the aleatoric and epistemic uncertainties in this notebook. Our analyses are now part of the notebook as well, however, we are a bit puzzled by the findings and struggled a bit with interpreting them. Would be happy to have your take on this :) @elcorto feel free to also chime in.

**Comment by nilsleh** [comment]
@codingS3b Thanks for the additional parts you added to the notebook. I am happy with merging it, and fixing the ruff error in a follow up PR, so no worries.

I really like how you go through the steps and build up the analysis along the way. And would also be interested in discussing the results you have, maybe a call is easier to do that?

**Comment by elcorto** [comment]
This PR is marked as draft because of several open points that we'd like to discuss:

* Explain more explicitly that we are (ab)using pixel regression to do multivariate regression by using a N x 1 pixel image. Since MVE produces a $\sigma_i, i=1,\ldots,N$ per pixel (output dimension), we also assume a diagonal covariance matrix, i.e. no correlations between output dimensions.

* Step 5 (Prediction): For some sample data points in the test set
  * plot predicted means of all ensemble members
  * plot epistemic and combined (epistemic and aleatoric) U as confidence bands to visualize there relative influence

* Step 6 (Calibration): At the moment there are open questions regarding the calculation and interpretation of calibration:
  * We do `plot_calibration(y_mean.flatten(), y_std.flatten(), y_test.flatten())` but don't explain why: because we assume a diagonal covariance, we also assume N independent 1D regression problems (one per output dimension) and calculate the average calibration over (i) all N 1D problems and (ii) all test set data points.
  * With the same assumption, we then investigate the calibration for each output dimension. Is this justified?
  * According to the box plots, the aleatoric U is highest and the epistemic lowest over the first 6 indices, where we would actually assume the highest epistemic U. For all other indices, the model only needs to learn the identity.
  * The per-dimension calibration plots don't seem to reflect the behavior of the aleatoric U as a function of the output dim found in the box plots, while the epistemic U seems to match.

**Comment by nilsleh** [comment]
@elcorto @codingS3b Thanks for the follow up.

I unfortunately have an upcoming deadline so not much time to currently dive into the notebook in depth, however, from the points you are making a couple of papers come to my mind that could possibly help interpret the results (maybe you know them already):

- Sources of uncertainty in Machine Learning [Gruber et al. 2023](https://arxiv.org/abs/2305.16703)
- Deep Ensembles Work but are they necessary [Abe et al. 2022](https://proceedings.neurips.cc/paper_files/paper/2022/file/da18c47118a2d09926346f33bebde9f4-Paper-Conference.pdf)
- Pathologies of Predictive Diversity in Deep Ensembles [Abe et al. 2024](https://arxiv.org/pdf/2302.00704)

From July 17 on, I will have some time to dive into the notebook myself. Thanks in advance, and looking forward to the discussion!

**Comment by nm19000** [comment]
>       * According to the box plots, the aleatoric U is highest and the epistemic lowest over the first 6 indices, where we would actually assume the highest epistemic U. For all other indices, the model only needs to learn the identity.
>       * The per-dimension calibration plots don't seem to reflect the behavior of the aleatoric U as a function of the output dim found in the box plots, while the epistemic U seems to match.

Concerning the aleatoric uncertainty, it should be noted that the NLL of a Gaussian is not numerically stable for sigma close to zero. Using this loss function when the data actually does not exhibit data uncertainty often leads to non-zero and larger sigmas. Due to this it may also be difficult to train the identity with the MVE model per dimension. Thus, also interpreting the later signal positions where there may be little aleatoric uncertainty is in itself a bit difficult.

**Comment by elcorto** [comment]
> Concerning the aleatoric uncertainty, it should be noted that the NLL of a Gaussian is not numerically stable for sigma close to zero. Using this loss function when the data actually does not exhibit data uncertainty often leads to non-zero and larger sigmas. Due to this it may also be difficult to train the identity with the MVE model per dimension. Thus, also interpreting the later signal positions where there may be little aleatoric uncertainty is in itself a bit difficult.

That is a very good point indeed, thanks. I think adding noise should resolve this, and would also be helpful from a didactic viewpoint.
